test/hotspot/gtest/opto/test_constant_division.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "runtime/os.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/growableArray.hpp"
+#include <type_traits>
+#include "unittest.hpp"
+
+template <class T, class UT>
+UT random();
+
+template <>
+juint random<jint, juint>() {
+  juint bits = juint(os::random()) % 31 + 1;
+  juint mask = (juint(1) << bits) - 1;
+  return os::random() & mask;
+}
+
+template <>
+juint random<juint, juint>() {
+  juint bits = juint(os::random()) % 32 + 1;
+  juint mask = bits == 32 ? std::numeric_limits<juint>::max() : (juint(1) << bits) - 1;
+  return os::random() & mask;
+}
+
+template <>
+julong random<jlong, julong>() {
+  juint bits = juint(os::random()) & 63 + 1;
+  julong mask = (julong(1) << bits) - 1;
+  julong full = (julong(os::random()) << 32) | os::random();
+  return full & mask;
+}
+
+template <>
+julong random<julong, julong>() {
+  juint bits = juint(os::random()) & 64 + 1;
+  julong mask = bits == 64 ? std::numeric_limits<julong>::max() : (julong(1) << bits) - 1;
+  julong full = (julong(os::random()) << 32) | os::random();
+  return full & mask;
+}
+
+template <class T>
+void magic_divide_constants(T d, T N_neg, T N_pos, juint min_s, T& c, bool& c_ovf, juint& s);
```

Comment:
I don't see any tests here of magic_divide_constants_round_down.

Reply:
I took the formula quite literally from the paper so I don't think there is a need for a separate test for those cases. It is also covered in the transformation tests from the Java side.

![image](https://github.com/openjdk/jdk/assets/49088128/cc1b1c5e-a37b-4510-9aff-839f0337a532)

Reply:
It is great that we are factoring out these algorithm steps into their own separately reviewable and maintainable API points.  I would even say it is *necessary* to do this, compared with the alternatives, such as (what we used to do) write random file-local helper functions or even hand-inlined statements.

A big part of the advantage is that we can catch bugs at the subroutine level, rather than at the system level.  But this only works if we write (or at least try hard to write) unit tests for each subroutine.  It doesn’t matter very much whether the subroutine comes from a published source or whether we created it ourselves somehow.  (The risk model for where failures come from differs a little, since the published source is presumably better reviewed than our own work.)  In either case a unit test (gtest) adds a lot of value.  The gtest can detect either of two interesting problems: 1. an error in the algorithm (this happens even if it is published) or 2. an error in our encoding of the algorithm.  The second is probably more likely.  It can happen due to source code errors or due to compiler bugs.  Either way, the gtest adds value by making it more likely that, if something goes wrong, we will find it before system integration and deployment.

So, based on what I can see here, I recommend writing at least a simple gtest for each subroutine we are writing, regardless of its source.

(Reminder:  When using random numbers as test inputs, please ensure that the gtests are seeded reproducibly.  Double check for pre-existing uses of random generators in the gtests that satisfy this requirement.)

Thanks for this very good work; I’m very glad you tackled it.  Getting numerics correct is always tricky, but this will pay off.

Reply:
@rose00 Thanks for your input, I have added a unit test for this case.